server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java

```diff
@@ -0,0 +1,182 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.cluster.routing.allocation.decider;
+
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.routing.RoutingNode;
+import org.opensearch.cluster.routing.ShardRouting;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.node.remotestore.RemoteStoreNodeService;
+import org.opensearch.node.remotestore.RemoteStoreNodeService.CompatibilityMode;
+import org.opensearch.node.remotestore.RemoteStoreNodeService.Direction;
+
+import java.util.Locale;
+
+/**
+ * A new allocation decider for migration of document replication clusters to remote store backed clusters:
+ * - For STRICT compatibility mode, the decision is always YES
+ * - For remote_store_enabled indices, relocation or allocation/relocation can only be towards a remote node
+ * - For "REMOTE_STORE" migration direction:
+ *      - New primary shards can only be allocated to a remote node
+ *      - New replica shards can be allocated to a remote node iff the primary has been migrated/allocated to a remote node
+ * - For other directions ("DOCREP", "NONE"), the decision is always YES
+ *
+ * @opensearch.internal
+ */
+public class RemoteStoreMigrationAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "remote_store_migration";
+
+    private Direction migrationDirection;
+    private CompatibilityMode compatibilityMode;
+    private boolean remoteStoreEnabledIndex = false;
+
+    public RemoteStoreMigrationAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.migrationDirection = RemoteStoreNodeService.MIGRATION_DIRECTION_SETTING.get(settings);
+        this.compatibilityMode = RemoteStoreNodeService.REMOTE_STORE_COMPATIBILITY_MODE_SETTING.get(settings);
+        this.remoteStoreEnabledIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(settings);
+        if (IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.exists(settings)) {
+            remoteStoreEnabledIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(settings);
+        }
+        clusterSettings.addSettingsUpdateConsumer(RemoteStoreNodeService.MIGRATION_DIRECTION_SETTING, this::setMigrationDirection);
+        clusterSettings.addSettingsUpdateConsumer(
+            RemoteStoreNodeService.REMOTE_STORE_COMPATIBILITY_MODE_SETTING,
+            this::setCompatibilityMode
+        );
+    }
+
+    // listen for changes in migration direction of cluster
```

Comment:
nit : this doesn't listen, but just sets the value . Same below

Reply:
Removed the comments, using self-explanatory function names.